### PR TITLE
Benchmarking guide

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

This PR adds a new benchmarking guide.

This closes https://github.com/NVIDIA/spark-rapids/issues/794

Note that this doc will need updating with instructions on using spark-submit to run the benchmarks once https://github.com/NVIDIA/spark-rapids/pull/904 is merged.

### CODE REVIEW

**Comment by andygrove** [comment]
build

File: docs/benchmarks.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+```
+import com.nvidia.spark.rapids.tests.tpcds._
+TpcdsLikeSpark.setupAllParquet(spark, "/path/to/tpcds")
+```
+
+The benchmark can be executed with the following syntax to execute the query and collect the 
+results to the driver.
+
+```
+TpcdsLikeBench.collect(spark, "q5", iterations=3)
```
Comment (tgravescs) [inline docs/benchmarks.md:76]: did we make tpch and tpcxbb all common like this?  I used to run tpch like: Q2Like(spark).collect()?

File: docs/benchmarks.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# Benchmarks
```
Comment (tgravescs) [inline docs/benchmarks.md:1]: should we have a link to this from integration test README and vice versa if it needs the setup stuff from that?

File: docs/benchmarks.md (hunk 1 above)
Comment (andygrove) [reply docs/benchmarks.md:76]: TPC-DS and TPCxBB are consistent so far. TPC-H has fallen behind a bit and I should probably update that too.

Comment (andygrove) [reply docs/benchmarks.md:76]: I have now updated https://github.com/NVIDIA/spark-rapids/pull/823 to make TPC-H consistent.

File: docs/benchmarks.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+Although it is possible to run benchmarks directly against the CSV data generated by the TPC data 
+generators, it is common to convert the data to Parquet format and run benchmarks against the 
+Parquet files instead.
+
+The `integration_test` module contains code for converting the CSV data sets to Parquet.
+
+The following commands can be entered into spark-shell to perform the conversion.
+
+```
```
Comment (sameerz) [inline docs/benchmarks.md:44]:
```suggestion
```scala
```

File: docs/benchmarks.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+The benchmarks can be executed in two modes currently:
+
+- Execute the query and collect the results to the driver
+- Execute the query and write the results to disk (in Parquet or CSV format)
+
+The following commands can be entered into spark-shell to register the data files that the 
+benchmark will query.
+
+```
```
Comment (sameerz) [inline docs/benchmarks.md:67]:
```suggestion
```scala
```

File: docs/benchmarks.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+import com.nvidia.spark.rapids.tests.tpcds._
+TpcdsLikeSpark.setupAllParquet(spark, "/path/to/tpcds")
+```
+
+The benchmark can be executed with the following syntax to execute the query and collect the 
+results to the driver.
+
+```
```
Comment (sameerz) [inline docs/benchmarks.md:75]:
```suggestion
```scala
```

File: docs/benchmarks.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+TpcdsLikeBench.collect(spark, "q5", iterations=3)
+```
+
+The benchmark can be executed with the following syntax to execute the query and write the results 
+to Parquet. There is also a `writeCsv` method for writing the output to CSV files.
+
+```
```
Comment (sameerz) [inline docs/benchmarks.md:82]:
```suggestion
```scala
```

File: docs/benchmarks.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+comparing between CPU and GPU benchmarks and a utility is provided to help with this.
+
+This is a simple utility that pulls results down to the driver for comparison so will only work for 
+data sets that can fit in the  driver's memory. 
+
+If data needs sorting before comparison, this is delegated to Spark before collecting the results.
+
+Example usage from spark-shell:
+
+```
```
Comment (sameerz) [inline docs/benchmarks.md:114]:
```suggestion
```scala
```

**Review by sameerz — COMMENTED** [review]
Add [info-strings](https://github.github.com/gfm/#info-string) to fenced code blocks

File: docs/benchmarks.md (hunk 1 above)
Comment (tgravescs) [reply docs/benchmarks.md:76]: great thanks

File: docs/benchmarks.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+For each of these benchmarks, source data must be generated by a utility that can generate data at
+any scale factor, where the scale factor is an integer representing approximately how many
+gigabytes of data will be generated, with scale factor 1 meaning ~1 GB, and scale factor 1000 meaning ~1 TB,
+for example.
+
+Further information on data generation can be found using the following links:
+
+- [TPC-DS Data Generator](https://github.com/databricks/tpcds-kit)
+- [TPC-H Data Generator](https://github.com/electrum/tpch-dbgen)
```
Comment (andygrove) [inline docs/benchmarks.md:19]: I don't know if this is the one we use for TPC-H or not. It was just the first result I saw in Google.

Comment (andygrove) [reply docs/benchmarks.md:19]: @revans2 Do you know if this is the correct TPC-H generator to link to here?

Comment (revans2) [reply docs/benchmarks.md:19]: This is what I used for TPC-H.  I have also downloaded some from http://www.tpc.org/tpch/ in the past, but you have to sign up for that.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

File: docs/benchmarks.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+```scala
+import com.nvidia.spark.rapids.tests.tpcds._
+TpcdsLikeSpark.csvToParquet(spark, "/path/to/input", "/path/to/output")
+```
+
+Note that the code for converting CSV to Parquet does not explicitly specify the number of 
+partitions to write, so the size of the resulting parquet files will vary depending on the value 
+for `spark.default.parallelism`, which by default is based on the number of available executor 
+cores. This value can be set explicitly to better control the size of the output files.
```
Comment (abellina) [inline docs/benchmarks.md:59]: oh nice you got to the bottom of this!

File: docs/benchmarks.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+It should also be noted that no decimal types will be output. The conversion code uses explicit 
+schemas to ensure that decimal types are converted to floating-point types instead.
+
+## Running Benchmarks
+
+The benchmarks can be executed in two modes currently:
+
+- Execute the query and collect the results to the driver
+- Execute the query and write the results to disk (in Parquet or CSV format)
```
Comment (abellina) [inline docs/benchmarks.md:69]: orc?

**Review by abellina — COMMENTED** [review]
had small comments, not blocking this to go in.

File: docs/benchmarks.md (hunk 10 above)
Comment (andygrove) [reply docs/benchmarks.md:69]: oops, yes ... I will fix this in the follow-up PR I am working on to update the docs to reflect other PRs that have been merged.

File: docs/benchmarks.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+TpcdsLikeSpark.csvToParquet(spark, "/path/to/input", "/path/to/output")
+```
+
+Note that the code for converting CSV to Parquet does not explicitly specify the number of 
+partitions to write, so the size of the resulting parquet files will vary depending on the value 
+for `spark.default.parallelism`, which by default is based on the number of available executor 
+cores. This value can be set explicitly to better control the size of the output files.
+
+It should also be noted that no decimal types will be output. The conversion code uses explicit 
+schemas to ensure that decimal types are converted to floating-point types instead.
```
[the remark this answers is not in the document]
Comment (andygrove) [reply docs/benchmarks.md:62]: Thanks, will fix this in the follow-up PR too.

**Pull request merged**